// sanity-plugin-vercel-deploy/dist/deploy-history.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _react = _interopRequireWildcard(require("react"));

var _axios = _interopRequireDefault(require("axios"));

var _spacetime = _interopRequireDefault(require("spacetime"));

var _ui = require("@sanity/ui");

var _icons = require("@sanity/icons");

var _deployStatus = _interopRequireDefault(require("./deploy-status"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var DeployHistory = function DeployHistory(_ref) {
  var _state$deployments;

  var url = _ref.url,
      vercelProject = _ref.vercelProject,
      vercelToken = _ref.vercelToken,
      vercelTeam = _ref.vercelTeam,
      hookContext = _ref.hookContext;

  var _useState = (0, _react.useState)({}),
      _useState2 = (0, _slicedToArray2["default"])(_useState, 2),
      state = _useState2[0],
      setState = _useState2[1];

  (0, _react.useEffect)(function () {
    if (!vercelProject) {
      return;
    }

    setState({
      loading: true
    });

    _axios["default"].get("https://api.vercel.com/v5/now/deployments?projectId=".concat(vercelProject, "&meta-deployHookId=").concat(url.split('/').pop(), "&limit=6").concat(vercelTeam !== null && vercelTeam !== void 0 && vercelTeam.id ? "&teamId=".concat(vercelTeam === null || vercelTeam === void 0 ? void 0 : vercelTeam.id) : ''), {
      headers: {
        'content-type': 'application/json',
        Authorization: "Bearer ".concat(vercelToken)
      }
    }).then(function (_ref2) {
      var data = _ref2.data;
      setState({
        deployments: data.deployments,
        loading: false,
        error: false
      });
    })["catch"](function (e) {
      setState({
        error: true,
        loading: false
      });
      console.warn(e);
    });
  }, [vercelProject]);

  if (state.loading) {
    return /*#__PURE__*/_react["default"].createElement(_ui.Flex, {
      direction: "column",
      align: "center",
      justify: "center",
      paddingTop: 3
    }, /*#__PURE__*/_react["default"].createElement(_ui.Spinner, {
      size: 4
    }), /*#__PURE__*/_react["default"].createElement(_ui.Box, {
      padding: 4
    }, /*#__PURE__*/_react["default"].createElement(_ui.Text, {
      size: 2
    }, "loading deployment history...")));
  }

  if (state.error) {
    return /*#__PURE__*/_react["default"].createElement(_ui.Card, {
      padding: 4,
      radius: 2,
      shadow: 1,
      tone: "critical"
    }, /*#__PURE__*/_react["default"].createElement(_ui.Text, {
      size: 2,
      align: "center"
    }, "Could not load deployments for ", vercelProject));
  }

  return /*#__PURE__*/_react["default"].createElement(_ui.Box, {
    as: 'ul',
    padding: 0
  }, /*#__PURE__*/_react["default"].createElement(_ui.Card, {
    as: 'li',
    padding: 4,
    borderBottom: true
  }, /*#__PURE__*/_react["default"].createElement(_ui.Flex, null, /*#__PURE__*/_react["default"].createElement(_ui.Box, {
    flex: 3
  }, /*#__PURE__*/_react["default"].createElement(_ui.Label, {
    muted: true
  }, "Preview URL")), /*#__PURE__*/_react["default"].createElement(_ui.Box, {
    flex: 1,
    marginLeft: 2
  }, /*#__PURE__*/_react["default"].createElement(_ui.Label, {
    muted: true
  }, "State")), /*#__PURE__*/_react["default"].createElement(_ui.Box, {
    flex: 3,
    marginLeft: 2,
    style: {
      maxWidth: '40%'
    }
  }, /*#__PURE__*/_react["default"].createElement(_ui.Label, {
    muted: true
  }, "Commit")), /*#__PURE__*/_react["default"].createElement(_ui.Box, {
    flex: 2,
    marginLeft: 2
  }, /*#__PURE__*/_react["default"].createElement(_ui.Label, {
    align: "right",
    muted: true
  }, "Deployed At")))), (_state$deployments = state.deployments) === null || _state$deployments === void 0 ? void 0 : _state$deployments.map(function (deployment) {
    var _deployment$meta, _deployment$meta2, _deployment$creator, _deployment$creator2, _deployment$creator3;

    return /*#__PURE__*/_react["default"].createElement(_ui.Card, {
      key: deployment.uid,
      as: 'li',
      padding: 4,
      borderBottom: true
    }, /*#__PURE__*/_react["default"].createElement(_ui.Flex, {
      align: "center"
    }, /*#__PURE__*/_react["default"].createElement(_ui.Box, {
      flex: 3
    }, /*#__PURE__*/_react["default"].createElement(_ui.Text, {
      weight: "semibold"
    }, /*#__PURE__*/_react["default"].createElement(_ui.Box, {
      style: {
        whiteSpace: 'nowrap',
        overflow: 'hidden',
        textOverflow: 'ellipsis'
      }
    }, /*#__PURE__*/_react["default"].createElement("a", {
      href: "https://".concat(deployment.url),
      target: "_blank",
      style: {
        color: 'inherit'
      }
    }, deployment.url)))), /*#__PURE__*/_react["default"].createElement(_ui.Box, {
      flex: 1,
      marginLeft: 2
    }, /*#__PURE__*/_react["default"].createElement(_ui.Text, null, /*#__PURE__*/_react["default"].createElement(_deployStatus["default"], {
      status: deployment.state
    }))), /*#__PURE__*/_react["default"].createElement(_ui.Box, {
      flex: 3,
      marginLeft: 2,
      style: {
        maxWidth: '40%'
      }
    }, /*#__PURE__*/_react["default"].createElement(_ui.Stack, {
      space: 2
    }, /*#__PURE__*/_react["default"].createElement(_ui.Text, null, /*#__PURE__*/_react["default"].createElement(_ui.Box, {
      style: {
        whiteSpace: 'nowrap',
        overflow: 'hidden',
        textOverflow: 'ellipsis'
      }
    }, (_deployment$meta = deployment.meta) === null || _deployment$meta === void 0 ? void 0 : _deployment$meta.githubCommitMessage)), /*#__PURE__*/_react["default"].createElement(_ui.Text, {
      size: 1,
      muted: true
    }, /*#__PURE__*/_react["default"].createElement(_ui.Inline, {
      space: 2
    }, /*#__PURE__*/_react["default"].createElement(_icons.TransferIcon, null), (_deployment$meta2 = deployment.meta) === null || _deployment$meta2 === void 0 ? void 0 : _deployment$meta2.githubCommitRef)))), /*#__PURE__*/_react["default"].createElement(_ui.Flex, {
      flex: 2,
      justify: "right",
      marginLeft: 2
    }, /*#__PURE__*/_react["default"].createElement(_ui.Inline, {
      space: 2
    }, /*#__PURE__*/_react["default"].createElement(_ui.Text, {
      style: {
        whiteSpace: 'nowrap'
      },
      muted: true
    }, _spacetime["default"].now().since((0, _spacetime["default"])(deployment.created)).rounded), /*#__PURE__*/_react["default"].createElement(_ui.Tooltip, {
      content: /*#__PURE__*/_react["default"].createElement(_ui.Box, {
        padding: 2
      }, /*#__PURE__*/_react["default"].createElement(_ui.Text, {
        muted: true,
        size: 1
      }, (_deployment$creator = deployment.creator) === null || _deployment$creator === void 0 ? void 0 : _deployment$creator.username)),
      fallbackPlacements: ['right', 'left'],
      placement: "top"
    }, /*#__PURE__*/_react["default"].createElement(_ui.Avatar, {
      alt: (_deployment$creator2 = deployment.creator) === null || _deployment$creator2 === void 0 ? void 0 : _deployment$creator2.username,
      src: "https://vercel.com/api/www/avatar/".concat((_deployment$creator3 = deployment.creator) === null || _deployment$creator3 === void 0 ? void 0 : _deployment$creator3.uid, "?&s=48"),
      size: 1
    }))))));
  }));
};

var _default = DeployHistory;
exports["default"] = _default;
