// sanity-plugin-vercel-deploy/dist/deploy-status.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _ui = require("@sanity/ui");

var _deployStatus = _interopRequireDefault(require("./deploy-status.css"));

var DeployStatus = function DeployStatus(_ref) {
  var status = _ref.status,
      justify = _ref.justify,
      children = _ref.children;
  return /*#__PURE__*/_react["default"].createElement(_ui.Flex, {
    wrap: "nowrap",
    align: "center",
    justify: justify,
    className: _deployStatus["default"].hookStatusIndicator,
    "data-indicator": status
  }, /*#__PURE__*/_react["default"].createElement(_ui.Box, {
    marginLeft: 2
  }, titleCase(status)), children);
};

var titleCase = function titleCase(str) {
  return str.toLowerCase().split(' ').map(function (word) {
    return word.charAt(0).toUpperCase() + word.slice(1);
  }).join(' ');
};

var _default = DeployStatus;
exports["default"] = _default;
