// sanity-plugin-vercel-deploy/dist/vercel-deploy.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _react = _interopRequireWildcard(require("react"));

var _nanoid = require("nanoid");

var _axios = _interopRequireDefault(require("axios"));

var _client = _interopRequireDefault(require("part:@sanity/base/client"));

var _components = require("@sanity/base/components");

var _ui = require("@sanity/ui");

var _icons = require("@sanity/icons");

var _deployItem = _interopRequireDefault(require("./deploy-item"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var initialDeploy = {
  title: '',
  project: '',
  team: '',
  url: '',
  token: ''
};

var VercelDeploy = function VercelDeploy() {
  var WEBHOOK_TYPE = 'webhook_deploy';
  var WEBHOOK_QUERY = "*[_type == \"".concat(WEBHOOK_TYPE, "\"] | order(_createdAt)");

  var client = _client["default"].withConfig({
    apiVersion: '2021-03-25'
  });

  var _useState = (0, _react.useState)(true),
      _useState2 = (0, _slicedToArray2["default"])(_useState, 2),
      isLoading = _useState2[0],
      setIsLoading = _useState2[1];

  var _useState3 = (0, _react.useState)(false),
      _useState4 = (0, _slicedToArray2["default"])(_useState3, 2),
      isSubmitting = _useState4[0],
      setIsSubmitting = _useState4[1];

  var _useState5 = (0, _react.useState)(false),
      _useState6 = (0, _slicedToArray2["default"])(_useState5, 2),
      isFormOpen = _useState6[0],
      setIsFormOpen = _useState6[1];

  var _useState7 = (0, _react.useState)([]),
      _useState8 = (0, _slicedToArray2["default"])(_useState7, 2),
      deploys = _useState8[0],
      setDeploys = _useState8[1];

  var _useState9 = (0, _react.useState)(initialDeploy),
      _useState10 = (0, _slicedToArray2["default"])(_useState9, 2),
      pendingDeploy = _useState10[0],
      setpendingDeploy = _useState10[1];

  var toast = (0, _ui.useToast)();

  var onSubmit = /*#__PURE__*/function () {
    var _ref = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee() {
      var vercelTeamID, vercelTeamName, _fetchTeam$data, fetchTeam;

      return _regenerator["default"].wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              // If we have a team slug, we'll have to get the associated teamId to include in every new request
              // Docs: https://vercel.com/docs/api#api-basics/authentication/accessing-resources-owned-by-a-team
              setIsSubmitting(true);

              if (!pendingDeploy.team) {
                _context.next = 18;
                break;
              }

              _context.prev = 2;
              _context.next = 5;
              return _axios["default"].get("https://api.vercel.com/v2/teams?slug=".concat(pendingDeploy.team), {
                headers: {
                  Authorization: "Bearer ".concat(pendingDeploy.token)
                }
              });

            case 5:
              fetchTeam = _context.sent;

              if (fetchTeam !== null && fetchTeam !== void 0 && (_fetchTeam$data = fetchTeam.data) !== null && _fetchTeam$data !== void 0 && _fetchTeam$data.id) {
                _context.next = 8;
                break;
              }

              throw new Error('No team id found');

            case 8:
              vercelTeamID = fetchTeam.data.id;
              vercelTeamName = fetchTeam.data.name;
              _context.next = 18;
              break;

            case 12:
              _context.prev = 12;
              _context.t0 = _context["catch"](2);
              console.error(_context.t0);
              setIsSubmitting(false);
              toast.push({
                status: 'error',
                title: 'No Team found!',
                closable: true,
                description: 'Make sure the token you provided is valid and that the team’s slug correspond to the one you see in Vercel'
              });
              return _context.abrupt("return");

            case 18:
              client.create({
                // Explicitly define an _id inside the vercel-deploy path to make sure it's not publicly accessible
                // This will protect users' tokens & project info. Read more: https://www.sanity.io/docs/ids
                _id: "vercel-deploy.".concat((0, _nanoid.nanoid)()),
                _type: WEBHOOK_TYPE,
                name: pendingDeploy.title,
                url: pendingDeploy.url,
                vercelProject: pendingDeploy.project,
                vercelTeam: {
                  slug: pendingDeploy.team || undefined,
                  name: vercelTeamName || undefined,
                  id: vercelTeamID || undefined
                },
                vercelToken: pendingDeploy.token
              }).then(function () {
                toast.push({
                  status: 'success',
                  title: 'Success!',
                  description: "Created Deployment: ".concat(pendingDeploy.title)
                });
                setIsFormOpen(false);
                setIsSubmitting(false);
                setpendingDeploy(initialDeploy); // Reset the pending webhook state
              });

            case 19:
            case "end":
              return _context.stop();
          }
        }
      }, _callee, null, [[2, 12]]);
    }));

    return function onSubmit() {
      return _ref.apply(this, arguments);
    };
  }(); // Fetch all existing webhooks and listen for newly created


  (0, _react.useEffect)(function () {
    var webhookSubscription;
    client.fetch(WEBHOOK_QUERY).then(function (w) {
      setDeploys(w);
      setIsLoading(false);
      webhookSubscription = client.listen(WEBHOOK_QUERY, {}, {
        includeResult: true
      }).subscribe(function (res) {
        var wasCreated = res.mutations.some(function (item) {
          return Object.prototype.hasOwnProperty.call(item, 'create');
        });
        var wasDeleted = res.mutations.some(function (item) {
          return Object.prototype.hasOwnProperty.call(item, 'delete');
        });

        if (wasCreated) {
          setDeploys(function (prevState) {
            return [].concat((0, _toConsumableArray2["default"])(prevState), [res.result]);
          });
        }

        if (wasDeleted) {
          setDeploys(function (prevState) {
            return prevState.filter(function (w) {
              return w._id !== res.documentId;
            });
          });
        }
      });
    });
    return function () {
      webhookSubscription && webhookSubscription.unsubscribe();
    };
  }, []);
  return /*#__PURE__*/_react["default"].createElement(_ui.ThemeProvider, {
    theme: _ui.studioTheme
  }, /*#__PURE__*/_react["default"].createElement(_ui.ToastProvider, null, /*#__PURE__*/_react["default"].createElement(_ui.Container, {
    display: "grid",
    width: 6,
    style: {
      minHeight: '100%'
    }
  }, /*#__PURE__*/_react["default"].createElement(_ui.Flex, {
    direction: "column"
  }, /*#__PURE__*/_react["default"].createElement(_ui.Card, {
    padding: 4,
    borderBottom: true
  }, /*#__PURE__*/_react["default"].createElement(_ui.Flex, {
    align: "center"
  }, /*#__PURE__*/_react["default"].createElement(_ui.Flex, {
    flex: 1,
    align: "center"
  }, /*#__PURE__*/_react["default"].createElement(_ui.Card, null, /*#__PURE__*/_react["default"].createElement("svg", {
    fill: "currentColor",
    viewBox: "0 0 512 512",
    height: "2rem",
    width: "2rem",
    xmlns: "http://www.w3.org/2000/svg"
  }, /*#__PURE__*/_react["default"].createElement("path", {
    d: "M256 48l240 416H16z"
  }))), /*#__PURE__*/_react["default"].createElement(_ui.Card, {
    marginX: 1,
    style: {
      opacity: 0.15
    }
  }, /*#__PURE__*/_react["default"].createElement("svg", {
    viewBox: "0 0 24 24",
    width: "32",
    height: "32",
    stroke: "currentColor",
    "stroke-width": "1",
    "stroke-linecap": "round",
    "stroke-linejoin": "round",
    fill: "none",
    "shape-rendering": "geometricPrecision"
  }, /*#__PURE__*/_react["default"].createElement("path", {
    d: "M16.88 3.549L7.12 20.451"
  }))), /*#__PURE__*/_react["default"].createElement(_ui.Card, null, /*#__PURE__*/_react["default"].createElement(_ui.Text, {
    as: "h1",
    size: 2,
    weight: "semibold"
  }, "Vercel Deployments"))), /*#__PURE__*/_react["default"].createElement(_ui.Box, null, /*#__PURE__*/_react["default"].createElement(_ui.Button, {
    type: "button",
    fontSize: 2,
    tone: "primary",
    padding: 3,
    radius: 3,
    text: "Add Project",
    onClick: function onClick() {
      return setIsFormOpen(true);
    }
  })))), /*#__PURE__*/_react["default"].createElement(_ui.Card, {
    flex: 1
  }, /*#__PURE__*/_react["default"].createElement(_ui.Stack, {
    as: 'ul'
  }, isLoading ? /*#__PURE__*/_react["default"].createElement(_ui.Card, {
    as: 'li',
    padding: 4
  }, /*#__PURE__*/_react["default"].createElement(_ui.Flex, {
    direction: "column",
    align: "center",
    justify: "center",
    paddingTop: 3
  }, /*#__PURE__*/_react["default"].createElement(_ui.Spinner, {
    size: 4
  }), /*#__PURE__*/_react["default"].createElement(_ui.Box, {
    padding: 4
  }, /*#__PURE__*/_react["default"].createElement(_ui.Text, {
    size: 2
  }, "loading your deployments...")))) : deploys.length ? deploys.map(function (deploy) {
    return /*#__PURE__*/_react["default"].createElement(_ui.Card, {
      key: deploy._id,
      as: 'li',
      padding: 4,
      borderBottom: true
    }, /*#__PURE__*/_react["default"].createElement(_deployItem["default"], {
      key: deploy._id,
      name: deploy.name,
      url: deploy.url,
      id: deploy._id,
      vercelProject: deploy.vercelProject,
      vercelTeam: deploy.vercelTeam,
      vercelToken: deploy.vercelToken
    }));
  }) : /*#__PURE__*/_react["default"].createElement(_ui.Card, {
    as: 'li',
    padding: 5,
    paddingTop: 6
  }, /*#__PURE__*/_react["default"].createElement(_ui.Flex, {
    direction: "column",
    align: "center",
    justify: "center"
  }, /*#__PURE__*/_react["default"].createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    fill: "none",
    width: "150",
    viewBox: "0 0 260 235"
  }, /*#__PURE__*/_react["default"].createElement("path", {
    fill: "white",
    fillRule: "evenodd",
    stroke: "black",
    strokeDasharray: "4 4",
    strokeWidth: "2",
    d: "M107.36 2.48l105.7 185.47H2.66L108.35 2.48z",
    clipRule: "evenodd"
  }), /*#__PURE__*/_react["default"].createElement("ellipse", {
    cx: "182.68",
    cy: "156.48",
    fill: "white",
    rx: "74.32",
    ry: "74.52"
  }), /*#__PURE__*/_react["default"].createElement("path", {
    stroke: "black",
    strokeWidth: "2",
    d: "M256.5 156.48c0 40.88-33.05 74.02-73.82 74.02-40.77 0-73.83-33.14-73.83-74.02 0-40.87 33.06-74.01 73.83-74.01 40.77 0 73.82 33.14 73.82 74.01z"
  }), /*#__PURE__*/_react["default"].createElement("mask", {
    id: "a",
    width: "149",
    height: "150",
    x: "108",
    y: "81",
    maskUnits: "userSpaceOnUse"
  }, /*#__PURE__*/_react["default"].createElement("ellipse", {
    cx: "182.68",
    cy: "156.48",
    fill: "#fff",
    rx: "74.32",
    ry: "74.52"
  })), /*#__PURE__*/_react["default"].createElement("g", {
    mask: "url(#a)"
  }, /*#__PURE__*/_react["default"].createElement("path", {
    fill: "black",
    fillRule: "evenodd",
    d: "M108.36 2.48l105.7 185.47H2.66L108.35 2.48z",
    clipRule: "evenodd"
  }))), /*#__PURE__*/_react["default"].createElement(_ui.Flex, {
    direction: "column",
    align: "center",
    padding: 4
  }, /*#__PURE__*/_react["default"].createElement(_ui.Text, {
    size: 2
  }, "No deployments created yet."), /*#__PURE__*/_react["default"].createElement(_ui.Box, {
    padding: 4
  }, /*#__PURE__*/_react["default"].createElement(_ui.Button, {
    fontSize: 3,
    paddingX: 5,
    paddingY: 4,
    tone: "primary",
    radius: 4,
    text: "Add Project",
    onClick: function onClick() {
      return setIsFormOpen(true);
    }
  })), /*#__PURE__*/_react["default"].createElement(_ui.Text, {
    size: 1,
    weight: "semibold",
    muted: true
  }, /*#__PURE__*/_react["default"].createElement("a", {
    href: "https://github.com/ndimatteo/sanity-plugin-vercel-deploy#your-first-vercel-deployment",
    target: "_blank",
    rel: "noopener noreferrer",
    style: {
      color: 'inherit'
    }
  }, "Need help?"))))))))), isFormOpen && /*#__PURE__*/_react["default"].createElement(_ui.Dialog, {
    header: "New Project Deployment",
    id: "create-webhook",
    width: 1,
    onClickOutside: function onClickOutside() {
      return setIsFormOpen(false);
    },
    onClose: function onClose() {
      return setIsFormOpen(false);
    },
    footer: /*#__PURE__*/_react["default"].createElement(_ui.Box, {
      padding: 3
    }, /*#__PURE__*/_react["default"].createElement(_ui.Grid, {
      columns: 2,
      gap: 3
    }, /*#__PURE__*/_react["default"].createElement(_ui.Button, {
      padding: 4,
      mode: "ghost",
      text: "Cancel",
      onClick: function onClick() {
        return setIsFormOpen(false);
      }
    }), /*#__PURE__*/_react["default"].createElement(_ui.Button, {
      padding: 4,
      text: "Create",
      tone: "primary",
      loading: isSubmitting,
      onClick: function onClick() {
        return onSubmit();
      },
      disabled: isSubmitting || !pendingDeploy.project || !pendingDeploy.url || !pendingDeploy.token
    })))
  }, /*#__PURE__*/_react["default"].createElement(_ui.Box, {
    padding: 4
  }, /*#__PURE__*/_react["default"].createElement(_ui.Stack, {
    space: 4
  }, /*#__PURE__*/_react["default"].createElement(_components.FormField, {
    title: "Display Title",
    description: "Give your deploy a name, like 'Production'"
  }, /*#__PURE__*/_react["default"].createElement(_ui.TextInput, {
    type: "text",
    value: pendingDeploy.title,
    onChange: function onChange(e) {
      e.persist();
      setpendingDeploy(function (prevState) {
        var _e$target;

        return _objectSpread(_objectSpread({}, prevState), {
          title: e === null || e === void 0 ? void 0 : (_e$target = e.target) === null || _e$target === void 0 ? void 0 : _e$target.value
        });
      });
    }
  })), /*#__PURE__*/_react["default"].createElement(_components.FormField, {
    title: "Vercel Project Name",
    description: "The exact name of the associated project on Vercel"
  }, /*#__PURE__*/_react["default"].createElement(_ui.TextInput, {
    type: "text",
    value: pendingDeploy.project,
    onChange: function onChange(e) {
      e.persist();
      setpendingDeploy(function (prevState) {
        var _e$target2;

        return _objectSpread(_objectSpread({}, prevState), {
          project: e === null || e === void 0 ? void 0 : (_e$target2 = e.target) === null || _e$target2 === void 0 ? void 0 : _e$target2.value
        });
      });
    }
  })), /*#__PURE__*/_react["default"].createElement(_components.FormField, {
    title: "Vercel Team Slug",
    description: "Required for projects under a Vercel Team (use team page URL slug)"
  }, /*#__PURE__*/_react["default"].createElement(_ui.TextInput, {
    type: "text",
    value: pendingDeploy.team,
    onChange: function onChange(e) {
      e.persist();
      setpendingDeploy(function (prevState) {
        var _e$target3;

        return _objectSpread(_objectSpread({}, prevState), {
          team: e === null || e === void 0 ? void 0 : (_e$target3 = e.target) === null || _e$target3 === void 0 ? void 0 : _e$target3.value
        });
      });
    }
  })), /*#__PURE__*/_react["default"].createElement(_components.FormField, {
    title: "Deploy Hook URL",
    description: "The Vercel deploy hook URL from your project's Git settings"
  }, /*#__PURE__*/_react["default"].createElement(_ui.TextInput, {
    type: "text",
    inputMode: "url",
    value: pendingDeploy.url,
    onChange: function onChange(e) {
      e.persist();
      setpendingDeploy(function (prevState) {
        var _e$target4;

        return _objectSpread(_objectSpread({}, prevState), {
          url: e === null || e === void 0 ? void 0 : (_e$target4 = e.target) === null || _e$target4 === void 0 ? void 0 : _e$target4.value
        });
      });
    }
  })), /*#__PURE__*/_react["default"].createElement(_components.FormField, {
    title: "Vercel Token",
    description: "A Vercel token from your account settings"
  }, /*#__PURE__*/_react["default"].createElement(_ui.TextInput, {
    type: "text",
    value: pendingDeploy.token,
    onChange: function onChange(e) {
      e.persist();
      setpendingDeploy(function (prevState) {
        var _e$target5;

        return _objectSpread(_objectSpread({}, prevState), {
          token: e === null || e === void 0 ? void 0 : (_e$target5 = e.target) === null || _e$target5 === void 0 ? void 0 : _e$target5.value
        });
      });
    }
  })), /*#__PURE__*/_react["default"].createElement(_ui.Card, {
    padding: 4,
    paddingBottom: 5,
    radius: 3,
    shadow: 1,
    tone: "caution"
  }, /*#__PURE__*/_react["default"].createElement(_ui.Box, {
    marginBottom: 2,
    style: {
      textAlign: 'center'
    }
  }, /*#__PURE__*/_react["default"].createElement(_ui.Inline, {
    space: 1
  }, /*#__PURE__*/_react["default"].createElement(_icons.WarningOutlineIcon, {
    style: {
      fontSize: 24
    }
  }), /*#__PURE__*/_react["default"].createElement(_ui.Heading, {
    size: 1
  }, "Careful!"))), /*#__PURE__*/_react["default"].createElement(_ui.Text, {
    size: 1,
    align: "center"
  }, "Once you create this deployment you will not be able to edit it.")))))));
};

var _default = VercelDeploy;
exports["default"] = _default;
